### PR TITLE
Threefold repetition detection

### PR DESCRIPTION
**Pull request by saproj**
Implement a threefold repetition detection. Below are the examples of problems fixed by this change.

1. Loosing move in a drawn position.
position fen 8/k7/3p4/p2P1p2/P2P1P2/8/8/K7 w - - 0 1 moves a1a2 a7a8 a2a1
The old code suggested a loosing move "bestmove a8a7", the new code suggests "bestmove a8b7" leading to a draw.

2. Incorrect evaluation (happened in a real game in TCEC Season 9).
position fen 4rbkr/1q3pp1/b3pn2/7p/1pN5/1P1BBP1P/P1R2QP1/3R2K1 w - - 5 31 moves e3d4 h8h6 d4e3
The old code evaluated it as "cp 0", the new code evaluation is around "cp -50" which is adequate.

Brings 0.5-1 ELO gain. Passes [-3.00,1.00].

STC: http://tests.stockfishchess.org/tests/view/584ece040ebc5903140c5aea
LLR: 2.96 (-2.94,2.94) [-3.00,1.00]
Total: 47744 W: 8537 L: 8461 D: 30746

LTC: http://tests.stockfishchess.org/tests/view/584f134d0ebc5903140c5b37
LLR: 2.96 (-2.94,2.94) [-3.00,1.00]
Total: 36775 W: 4739 L: 4639 D: 27397

### CODE REVIEW

**Comment by MichaelB7** [comment]
A nice effort with a satisfactory result. Congrats.

**Comment by Stefano80** [comment]
Why -3, 1? It adds code, so it should be 0, 5

**Comment by saproj** [comment]
Can be considered a bug fix, therefore tested for no-regression.

**Comment by joergoster** [comment]
@Stefano80 Similar tries were also tested with bugfix bounds in the past. But I admit this patch adds a LOT of code. OTOH, this is the first one that passed STC and LTC!
I don't envy the maintainer(s) for the decision they have to make.
@saproj Congrats!

**Comment by atumanian** [comment]
Good work! The 3-fold repetition blindness has been always annoying.
But why does this change the bench signature? The bench test doesn't have pre-root positions, so the patch shouldn't affect it.

**Comment by saproj** [comment]
Because the root position itself is reenterable unless it is not a repetition (like in the case of bench). Why? I tested the alternative and it was weaker.

**Comment by syzygy1** [comment]
It does not seem to make sense to allow a first repetition of the root position. Did keeping the old behaviour for the root position fail on fishtest?

A possible way to remove the st->draw flags: after setting up the position in uci.cpp, walk from the root back into history and set all non-repeated st->key values to zero.

**Comment by saproj** [comment]
> It does not seem to make sense to allow a first repetition of the root position. Did keeping the old behaviour for the root position fail on fishtest?

Yes, it did.

> A possible way to remove the st->draw flags: after setting up the position in uci.cpp, walk from the root back into history and set all non-repeated st->key values to zero.

Interesting idea. I'll try it.

**Comment by vdbergh** [comment]
@saproj The fact that the first version failed SPRT(-3,1) does not mean it is weaker. You cannot draw any conclusion from a failed SPRT(-3,1). 

Also in case of a small regression (like -1 elo as could be the case here) you can make a SPRT(-3,1) pass by repeating it enough times (possibly combining it with placebo changes a la "Take 1", "Take 2", "optimized version" etc...). 

I think in this case the maintainers should simply be ready to (possibly) sacrifice 1 elo in order to repair what is obviously a bug in Stockfish.

**Comment by MichaelB7** [comment]
I know I don't have a vote - but I would vote for this patch.  There are many users of Stockfish that would appreciate the nuance corrected here.

**Comment by Stefano80** [comment]
I really don't understand why it should be tested with [-3, 1]. It it adds more than 30 lines of code. It does not fix anything, and for certain not a bug. It has not been approved in advance by maintainers. It is just an arbitrary abuse of the framework. The guidelines say:

> These tests are also used for bug fixes and other special cases, but only after being discussed and approved in advance to avoid people testing with no-regression mode becoming their preferred toy, instead of using the stricter standard mode.

Please either stick to the guidelines or propose to change them.

@vdbergh : it has been repeated hundreds of times that conter-intuitive eval in an arbitrary class of positions is not considered a bug.

**Comment by Mindbreaker1** [comment]
I think 30 lines is an exaggeration. Some is documentation, spacing, and brackets.  I see 13 without that.  It fixes a problem and it does not appear to hurt.  And is it possible that this might help when contempt is present achieving more at a lower setting?  That is just a guess, not an assertion.

At the very least, it means less grumbles on the forum.  That has to be worth something.

**Comment by saproj** [comment]
@vdbergh I did direct comparison locally. Not allowing to reenter the root position is, surprisingly, worse  than allowing it. By several ELO points. Why? It's a mystery.

**Comment by saproj** [comment]
@Stefano80 Just to clarify my position: I believe I'm fixing a bug. Surprised by your "for certain not a bug".

**Comment by ddugovic** [comment]
Regardless we can agree, "It has not been approved in advance by maintainers."

> The old code evaluated it as "cp 0", the new code evaluation is around "cp -50" which is adequate.

How is "cp 0" a bug?  (In my fork I do address this issue, however I cannot prove "cp 0" is a bug according to terminology used by maintainers.)

**Comment by jhellis3** [comment]
To understand why this is a "bug" is fairly straightforward - the current behavior simply does not adhere to the actual rules of chess.    Using a small eval difference is a bit disingenuous.  Say the eval is -10, but human players repeated the previous position.  Showing an eval of draw is clearly in error, since the first repetition does not result in an immediate draw under the actual rules of the game.  While it may not make any difference for Stockfish in playing a game, it does cause a clear discontinuity in evaluation when used for analysis.  As analysis is the primary use case for the engine, ensuring correct output should at least be worthy of consideration, IMHO.

**Comment by saproj** [comment]
@syzygy1 Zeroing non-repeated st->key (at least the way I implemented it) is not better. See http://tests.stockfishchess.org/tests/view/58514cd80ebc5903140c5c04

**Comment by saproj** [comment]
@ddugovic,
"cp 0" is a bug because it is a wrong evaluation. This 0 eval may be assigned to a position which is in fact very different from equality.

You mentioned you address this issue in your fork. I do not know about it. So, is there also a fix for the 1st problem from my list? I mean the problem when SF makes a loosing move in a drawn position.

**Comment by ddugovic** [comment]
[My fork's change](http://github.com/ddugovic/Stockfish/commit/fb16fb4ff946f8456c1eded29d903b1273e16b78) probably looks familiar to some; unfortunately I forget who the original author is that I copied the idea from.

People keep using the words "bug" and "fix" and while I only endorse the word "problem" my fork does not play `a8a7` in that study position.

**Comment by saproj** [comment]
@ddugovic Familiar idea. It introduces an ELO-stealing slowdown because of additional loop iterations.

**Comment by Rocky640** [comment]
Congrats to the author !!!

There is one helper function, and a few line changes. 
This looks quite reasonable to me.

Let me take 2 examples:
Let's say that engine would systematically output cp=0 whenever the last move was e.p.
I think we would  want to fix this.

If search would never consider promotions to bishop, (hello old Rybka),
I think we would  want to fix this. No ELO improvement,
but we want to design an engine which plays according to the rules of chess.

If I have a car which runs perfectly, but would stop working each time I "leave from work" (but would be ok if I leave from all the other places), I would say that this is a problem.

As it is, each time current master "leaves from a repetition", it virtually stops doing its duty which is 
to keep trying to output the most realistic evaluation according to the available information.

It's not like a specific funny position which is misevaluated. This is a systemic error which does not really depend of the position. It depends only to the fact that there was a repetition in the game history.

We expect an engine to properly handle e.p., castling rights, 50 moves rule, (which all relates to the game history) AND also repetitions.

It might even be a small ELO gainer. In a time limited game, SF might go for a first repetition, and on the next move. using the info in the TT, and with the time allocated by the time management, it might go for a better fighting continuation or simply keep trying with something else.

**Comment by lantonov** [comment]
Having tried myself to fix this problem (bug?) and failing badly, I know how complicated it is if you dig deeper. The author of this patch did a good job, improving on his previous idea.

**Comment by syzygy1** [comment]
@saproj 
I would not expect zeroing st->key instead of adding st->draw flags to do measurably better, but it cannot do worse and, arguably, it is a slightly simpler solution (not needing an additional field in StateInfo, for example).

**Comment by atumanian** [comment]
@syzygy1, but your version treats the root position differently: it doesn't allow repeating it even once during a search, unlike saproj's version. He has already written about this here.

**Comment by saproj** [comment]
@syzygy1 This idea of zeroing st->key looked appealing to me at first. But now that I have tested it, I want to stick to adding st->draw. By the way, this new field resides in an alignment gap (under x86_64) and StateInfo's size did not change.

**Comment by ElbertoOne** [comment]
I'm in favor of this patch, even if it adds some complexity. To remove all doubt, we might consider a reverse simplification test (master against patch instead of patch versus master).

**Comment by jcalovski** [comment]
If I were a maintainer i'd list a few "exceptions":
- 3 fold rep
- fortress detection
- reworked contempt
- and so on...
For which i'd relax the rules slightly.

**Comment by stockfishdeveloper** [comment]
I guess this patch cuts to the root of Stockfish's purpose.
The question it brings up is:
Should Stockfish be made for playing or analyzing?

**Comment by jhellis3** [comment]
Well, when was the last even odds (regular) match between a human being and Stockfish?

**Comment by ddugovic** [comment]
When a person selects which engine to download, they consider both price (free) and rating.  It's not as if Stockfish has retired from competition.

**Comment by jhellis3** [comment]
And this patch neither affects the price nor the rating.... Also did not answer the question....

**Comment by ddugovic** [comment]
A direct answer for your question can be found at http://lichess.org/tv/computer .

**Comment by jhellis3** [comment]
Lol...

**Comment by syzygy1** [comment]
@atumanian 
Sure, but repeating the root is supposed to make no sense, for the same reason why repeating any other position after the root (so in the search tree) makes no sense. If starting from the root the search finds no better than a path returning to the root, the position should be scored a draw. This is not necessarily the case if you repeat a position coming before the root, because one of the moves that were played might be suboptimal.

Of course if it is confirmed that allowing repetitions of the root gives Elo, then that ends the discussion. But I doubt that it really does.

**Comment by ddugovic** [comment]
It took a couple examples to understand, and @syzygy1 is correct: supposing there is a better move than repeating, repeating the root (or repeating any position in the search tree) is a costly distraction.  But surely it's at least as costly to prevent such repetitions...

**Comment by cuddlestmonkey** [comment]
Personally, I run a locally-built SF which has the current "best" 3-fold patch merged in, because the display of cp0 when analysing human games is annoying. Not the end of the world when you are aware of what's happening and why, but annoying nonetheless. This patch, which appears to cost zero or negligible Elo, is the new champ for me. 

I understand why the maintainers might be loath to accept the patch for fear of opening the gates to other special pleadings. I also understand that this is the one issue that confuses users over and over again, and can't be kicked upstairs to be sorted out by the GUI (unlike the 'bad FEN' issues that crop up from time to time)

Perhaps we should simply, when releasing official binaries, release one for engine-engine play and one for analysis.

**Comment by atumanian** [comment]
@syzygy1, I agree with you. It's a mystery why allowing to repeat the root position is beneficial for strength.
Surprisingly this makes sense when MultiPV > 1, because if a player can force returning to the root position with a move then this move is one of the best.

**Comment by atumanian** [comment]
@cuddlestmonkey, tests show that the new version isn't weaker, so there's no problem applying the patch.

**Comment by syzygy1** [comment]
@atumanian 
You certainly make an interesting point about MultiPV > 1.

If the _best_ move leads to a repeat of the root position, then there is no reason to score the move better than a draw (as far as I can see).

But if the Nth best move with N > 1 leads to a repetition of the root position, then you can still play the best move from there. So one could argue that the value of this Nth best move is equal to the value of the best move.

But I don't know if this is a good reason to allow repeats of the root position in case of MultiPV > 1. I don't think the user actually wants to see those repeating Nth best moves scored as high as the best move. And in practice, letting those moves repeat the root position in their search and then choose the best move as continuation will result in a much shallower search and perhaps a score that is even higher than that of the "real" best move.

So I _think_ that also for the purpose of MultiPV it is still preferable to score the first repetition of the root position as draw. Showing a draw score for a suboptimal move that only wins by repeating the root position and then playing the best move seems acceptable.

**Comment by joergoster** [comment]
@syzygy1 Well, in human games this happens quite often. Why? To step over the next time control!
As we almost exclusively test with fisher tc, it would be interesting to test this patch with repeating tc, too.

But then it definitely makes no sense to me to already evaluate the first repetition as a draw ... :-)

**Comment by syzygy1** [comment]
@joergoster 
The first repetition of the root position should be evaluated as a draw not because chess has a draw by repetition rule, but because it makes no sense to award a positive score to running in circles (nor a negative score -> if a first repetition can be forced, then second, third, fourth, fifth repetitions can be forced too).

The MultiPV > 1 thing is about the case where progress _can_ be made with the best root move, but some other moves only lead (at best) to a repetition of the root position after which the best root move can still be played. Sure, a human player might decide to play such a suboptimal move just to gain some time on the clock, but MultiPV > 1 is not meant for that. And teaching SF to go for second best repeating moves to gain time on the clock is probably 1000x as difficult to achieve without losing Elo as it is to teach SF to play easy moves without losing Elo (already difficult enough). Merely allowing repeats of the root position is certainly not going to help anything in this respect.

**Comment by syzygy1** [comment]
@ddugovic 
What do you mean by "it's at least as costly to prevent such repetitions"?
Scoring the first repetition of the root position as a draw is not costly to implement.

Are you thinking of tree size?

**Comment by ddugovic** [comment]
I'm thinking that repetition of the root position seems an uncommon occurrence, and curious whether re-entry of the root position could have benefits (as opposed to entry of positions where either color "marks time" to avoid repetition).

**Comment by mcostalba** [comment]
Is there someone (possibly the author) that can explain what the patch does in details?

I have looked at the posts but there is no reference to what the patch does, Thanks.

**Comment by atumanian** [comment]
@syzygy1, MultiPV > 1 is used only for analysis, not for playing. It must show several best moves with their scores. If a move leads to a repetition of the root position the move is not suboptimal but optimal, and should be score as such.

**Comment by atumanian** [comment]
@mcostalba, the patch implements the 3-fold repetition rule.
In the current version, once a position which has repeated once earlier occurs during a search, it is immediately scored as a draw. In the new version, it's scored with the following operation. If the position repeats once earlier but strictly after the root, or repeats twice before or at the root, the position is scored as a draw. Otherwise, the search continues as normal.

**Comment by mcostalba** [comment]
Thanks, but why drawDepth instead of pliesfromnull?

**Comment by atumanian** [comment]
Because drawDepth is reset at null moves as well as at pawn moves and captures.

**Comment by mcostalba** [comment]
Also previous code, of course.

**Comment by mcostalba** [comment]
@saproj can you please revert back to use pliesfromnull and retire drawDepth? 

I would like to see just what is really need, i. e the functionality change of your patch.

File: src/position.cpp [hunk 1]
```diff
@@ -839,6 +839,9 @@ void Position::do_move(Move m, StateInfo& newSt, bool givesCheck) {
   // Set capture piece
   st->capturedPiece = captured;
 
+  // Draw at first repetition
+  st->draw = true;
```
Comment (mcostalba) [inline src/position.cpp]: I don't like to set st->draw = true; for each searched node in do_move(), this seems to me a real waste.

Eventually you can use gamePly in the draw condition like:

       if (stp->key == st->key)
           return stp->gamePly > RootPosition->gamePly || stp->draw;

**Comment by syzygy1** [comment]
@mcostalba 
The alternative implementation that I proposed does not need an st->draw field and requires no change in do_move() or in is_draw().

Instead, after setting up a position in uci.cpp, the st->key field is set to 0 for all history positions that do not repeat earlier history positions. (By history position I mean the game positions that came before the root position.)

The effect is that is_draw() detects a draw
- if the current position repeats a position from the root up to the current position, or
- if the current positions repeats a _repeated_ history position (so a position that was on the board before the root position).

**Comment by syzygy1** [comment]
@atumanian 
My MultiPV > 1 comment was written having in mind that MultiPV > 1 is purely intended for analysis.

Suppose there is one single best move that makes progress and there are four moves that are only "good" because they will repeat the root position after which the single best move can be played. Do you really consider it desirable to display all five moves with the same score as if they are really equally good? I do not think that is useful or that that is what the analysing user would want. On the contrary, I think such moves are clearly inferior to the single best move and ideally should be displayed as being inferior. (Look at it this way: the second time the root position occurs, those same 4 moves in that same position clearly aren't so good anymore at all. So why would they have been good the first time around?)

And even worse, since the repeat of the root position has a delaying effect (decreasing the effective depth at which the move is searched), chances are that those some or all of those four moves will be displayed with _higher_ scores than the single move that makes progress. This happens when repeating the root position allows the search to move some threat beyond the search horizon.

Of course the latter observation applies also to MultiPV = 1. The move repeating the move position might score better than the single best move due to the delaying effect, resulting in an unnecessary repetition of the root position.

File: src/position.cpp (hunk 1 above)
Comment (saproj) [inline src/position.cpp]: Well, if I find a way to access the root position from Position::is_draw() I will create a fishtest test to compare the two approaches. I already did that for syzygy's idea and it turned out to be not good :).

**Comment by mcostalba** [comment]
Is this idea functional equivalent to the proposed patch?

The patch author said your idea has been tested and failed. If your idea is
functional equivalent then should fail also current patch.

**Comment by syzygy1** [comment]
@mcostalba 
The only functional difference is that saproj's implementation allows one repetition of the root position. According to him, not allowing first repetitions of the root position hurts Elo.

As far as I can see there is no reason why repeating the root position would ever be good, so if sapproj's tests suggest otherwise then I am blaming that on statistical noise. (Also it is dangerous and in fact wrong to draw conclusions like 0.5-1 Elo gain from a [-3,1] test, see the top comment. I haven't checked all the tests that were run, but I don't think it can be statistically soundly concluded that scoring the first repetition of the root position is bad.)

Note that current SF does not allow any repetion (i.e. any first repetition is immediately scored as draw). So it does not allow repetitions of the root position either.

The "problem" of showing incorrect draw scores is caused by not allowing first repetitions of positions that come before the root position.

Allowing repetition of the root position and any position coming after it (in the search tree) is just a waste of effort. If the "best" path leads back to the root, then the position is  draw. If there is any way at all to improve starting from the root position, then that can be done without repeating the root position.

**Comment by mcostalba** [comment]
Can't you zero out also root position key?

Anyhow intuitively I also agree with you that allowing to recourse in root
position has no valid explanation so far.

BTW the best approach IMO is to pass depth argument to is_draw() function :
in this way you don't need tricks and don't need to alter the meaning of
position keys just for a subset of positions.

**Comment by syzygy1** [comment]
Btw, it was always my understanding that fixing this problem in Stockfish hurt Elo (at least in self play) not because correctly testing for repetitions is expensive (it does not need to be expensive at all), but because the fix increases the size of the search tree.

It may be that previous attempts were less efficiently implemented than this attempt, but the speed difference should still be very small.

It may be that allowing first repeats of the root position is the magic trick that makes it work, but I have difficult believing that. (I may of course turn out to be wrong on this. I just don't expect it.)

**Comment by syzygy1** [comment]
@mcostalba 
I guess zeroing the root position key gives a problem when making moves from the root position.
This can of course be fixed, but it is messy. (More importantly, it seems wrong to me... of course tests may prove me wrong.)

I don't know how passing depth would work here. (depth is kind of unrelated to the number of plies to the root).

I think the only reason to keep around StateInfo objects for the "history" positions is for repetition detection, so messing a little bit with those objects (i.e. setting st->key = 0 for those you don't want to count as repetition) seems acceptable.

**Comment by atumanian** [comment]
@syzygy1 

> Suppose there is one single best move that makes progress and there are four moves that are only "good" because they will repeat the root position after which the single best move can be played. Do you really consider it desirable to display all five moves with the same score as if they are really equally good? I do not think that is useful or that that is what the analysing user would want. On the contrary, I think such moves are clearly inferior to the single best move and ideally should be displayed as being inferior. (Look at it this way: the second time the root position occurs, those same 4 moves in that same position clearly aren't so good anymore at all. So why would they have been good the first time around?)

Yes, it's desirable to display the 4 moves as equally good. Score is an estimate of whether the moves leads to a win, a draw or a loss with the optimal play. It doesn't matter how many moves the game will last. For example, a move which leads to a mate in 50 moves is as good as a move which mates immediately.
It's a well-known fact that the score of a move depend on the history of the position. I don't understand why this seems strange for you.

> 
> And even worse, since the repeat of the root position has a delaying effect (decreasing the effective depth at which the move is searched), chances are that those some or all of those four moves will be displayed with higher scores than the single move that makes progress. This happens when repeating the root position allows the search to move some threat beyond the search horizon.
> 
> Of course the latter observation applies also to MultiPV = 1. The move repeating the move position might score better than the single best move due to the delaying effect, resulting in an unnecessary repetition of the root position.
Why is this bad if the engine goes for a repetition? It may be actually the best choice. For example, the engine has figured out that with a move it can force repeating the root position, but has pruned some of its subsequent moves. What is the best choice? Of course, playing the move which leads to a repetition! It won't be worse, but later the engine may find a better line out of the loop!

**Comment by syzygy1** [comment]
@atumanian 
There is nothing strange here, I understand this stuff quite well. And I am pretty much convinced that a serious chess player will not consider a move that merely repeats the position to be as good as the move that makes progress. This seems terribly obvious to me. That "game-theoretically" the repeating move might be "as good" is nice for game theoreticians, but that does not help for analysis.

If you don't see why going for a repetition **of the root position** is never the best choice unless the engine thinks the position is a draw, then think again. A PV that halfway repeats the root position is an artificially shortened PV that apparently was preferred over the "real" PV because it hides something bad behind the horizon.

(Things are entirely different if we are talking about repeating a position that comes before the root position.)

And sure, the engine might still find something better before actually repeating the root position, **just like the opponent might find something better** before the engine manages to repeat the root position. Speculating that you have missed something is never a good idea.

I suppose you also see nothing bad in repeating positions within the search tree? Because all your arguments work just as well for that case.

The point is this: there is a crucial difference between allowing (first) repetitions of positions in the game history and allowing (first) repetitions of positions in the search tree. And in this respect, the root position is part of the search tree.

**Comment by mcostalba** [comment]
@syzygy1 yes depth is not the correct parameter, ply is. Something like (NOT TESTED): 

```
bool Position::is_draw(int ply) const {

  if (st->rule50 > 99 && (!checkers() || MoveList<LEGAL>(*this).size()))
      return true;

  int e = std::min(st->rule50, st->pliesFromNull);

  if (e < 4)
    return false;

  int d = e + 2;
  StateInfo* stp = st->previous->previous;

  do {
      stp = stp->previous->previous;

      if (stp->key == st->key)
          return d - e < ply || stp->draw;

  } while ((e -= 2) >= 4);

  return false;
}

```

You call it with 

`pos.is_draw(ss->ply)`

**Comment by ddugovic** [comment]
> It may be that allowing first repeats of the root position is the magic trick that makes it work, but I have difficult believing that.

> And I am pretty much convinced that a serious chess player will not consider a move that merely repeats the position to be as good as the move that makes progress... there is a crucial difference between allowing (first) repetitions of positions in the game history and allowing (first) repetitions of positions in the search tree. And in this respect, the root position is part of the search tree.

From a player's perspective I'm not sure it's quite that simple, at least not until progress is found!  Supposing there is a magic trick involving the root position, this might be how it works under the "attacker" circumstance?

* A defender (whose position is worse) is pragmatic to evaluate any search tree repetition as "draw" (or "lower-bound draw").

* An attacker who finds a search tree first repetition could be pragmatic as @atumanian suggests to not evaluate it as "draw" but instead allow expansion of it (with many TT hits).

**Comment by atumanian** [comment]
@syzygy1 
> There is nothing strange here, I understand this stuff quite well. And I am pretty much convinced that a serious chess player will not consider a move that merely repeats the position to be as good as the move that makes progress. This seems terribly obvious to me. That "game-theoretically" the repeating move might be "as good" is nice for game theoreticians, but that does not help for analysis. Ask any player whether he thinks a move that mates in 50 is as good as a mate in 1 in the same position.

Why doesn't that help for analysis? For example, if a player played a moves that leads to a repetition and won the game, but the analyzing engine shows the move as a draw, does it help with analysis better?

**Comment by ddugovic** [comment]
> Why doesn't that help for analysis? For example, if a player played a move that leads to a repetition and won the game, but the analyzing engine shows the move as a draw, does it help with analysis better?

@atumanian Even today [players repeatedly report a 0.0 evaluation as not helpful](http://lichess.org/forum/lichess-feedback/engine-bug#1)!

**Comment by mcostalba** [comment]
I have rewritten the patch in away that seems more reasonable to me.

https://github.com/mcostalba/Stockfish/compare/master...3fold

**Comment by joergoster** [comment]
@mcostalba Nice! But this again excludes the root position to be repeated once without directly assigning a draw score ...

**Comment by vdbergh** [comment]
@mcostalba  If I read correctly this is precisely the version that failed several times in the past. It seemed to be a -1 elo regression. So if you keep on testing it, it will eventually pass... which would be a good thing IMHO.

**Comment by mcostalba** [comment]
Can you please post some link to validate your statement or you just go by
memory.

FYI this is fully equivalent to the original patch but the recursion at
root.

**Comment by vdbergh** [comment]
@mcostalba AFAICS your version does not contain the optimization "calc_draw". This is probably a placebo but one cannot be sure.

**Comment by atumanian** [comment]
This tests shows that it's better to allow repetition of the root position: http://tests.stockfishchess.org/tests/view/58514cd80ebc5903140c5c04
@syzygy1, I understand the difference between repeating the root position and repeating positions after the root. My only argument for repeating the root is the results of the tests.

**Comment by vdbergh** [comment]
@mcostalba As I said your "new" version has been tested many times before. Here is one example https://github.com/saproj/Stockfish/commit/1c5996e6239b8f39a06efb633721d83aa020c6b3

**Comment by vdbergh** [comment]
@mcostalba This seems to be the first version. It failed SPRT(-4,0). 

https://groups.google.com/forum/?fromgroups=#!searchin/fishcooking_results/3fold_fix%7Csort:relevance/fishcooking_results/Jvg9Ytp5mY4/JyQ7uviPMkIJ

**Comment by mcostalba** [comment]
These version are similar but not equivalent, in particular these have not
verify that the boundary is exacty at root position (it is tricky to find
the exact boundary, I did this by some debug code. You can see it in
previous commit of my pushed branch).

**Comment by mcostalba** [comment]
@vdbergh regarding the calc_draw(), well if thepatch passed just ebcause of that then it is better to fail and do not commit, it means it passed by pure luck. In real games teh cases where we reach plies before root are very very few (measured with debug instrumentation) and even inn those cases, the extra search to reach the game ply base is absolutely unmeasurable. Again, if this is teh case then it is much more probable to explain teh succesful test by a statistical fluke.

**Comment by vdbergh** [comment]
@mcostalba Thanks for repeating my points (your last post). As for your earlier post. I think the old version from 2014 is equivalent to your "new" one (there is nothing tricky about this). I have not checked it carefully however (it was not my patch).

**Comment by vdbergh** [comment]
BTW I was simply responding to your claim: "FYI this is fully equivalent to the original patch but the recursion at root.". This statement is false as the (probably placebo) optimization is dropped. Also: please do not use "FYI". I am perfectly capable of reading your code.

**Comment by mcostalba** [comment]
Ok, I have submitted a couple of test: with root position included and excluded from extended draw repetition.

**Comment by ddugovic** [comment]
I apologize for the untimeliness of the following idea as I do not fully understand [HGM's comment](http://en.lichess.org/forum/lichess-feedback/stockfish-only-requires-twofold-repeatition-to-draw?page=2#11) as it applies to Stockfish:

> One way to solve [threefold repetition] is to remove all positions from the game history that occurred only once, during loading of a position in AnalyzeMode. This is code that is unlikely to be patched very frequently. E.g. you could spoil the incrementally updated hash key before parsing the moves by XOR'ing it with some constant, and repair it afterwards. And repair the key of any encountered position that would get stored in the move-history table whenever you detect it is a repetition.

Given that the `position ... moves` command reconstructs the entire game history I wonder, after that command would it be safe (for positions only existing once in the game history) to spoil [stp->key](http://github.com/official-stockfish/Stockfish/blob/master/src/position.cpp#L1097) (with or without "repair" code)?

**Comment by atumanian** [comment]
@ddugovic, I've actually tried this idea on Fishtest: http://tests.stockfishchess.org/tests/view/58223a4b0ebc5910626b9c64
My version isn't perfect, so you shouldn't take it literally, but you can look at it.
Before every search it reconstructs pre-root states so that all positions which are repeated only once are deleted. The patch adds a lot of code, but the code which tests for a draw is almost unchanged.

**Comment by mcostalba** [comment]
Bith tests passed, so I think I will commit the one with root position excluded from extended draw repetition becuase it seems the most logical to me.

**Comment by atumanian** [comment]
@mcostalba, but only the version with the root position included works correctly with MultiPV > 1, as I have said earlier.

**Comment by atumanian** [comment]
We can also make a match between the two versions to decide which is stronger.

**Comment by mcostalba** [comment]
@atumanian The argument on MultiPV is not clear, in particular has been tentatively refuted by @syzygy1 and I think his arguments make sense. Overall to include root because of MultiPV leads to a questionable and subjective argumentation for both parts, not to a clear black or white condition. so because it is not clear the benefit in MultiPV case and because, apart form this very specific and debatable point, there is no logical/measurable added benefit, I will commit the version with root excluded.

**Comment by mcostalba** [comment]
Merged with 881a9dfb0a8fec3b1

**Comment by atumanian** [comment]
@mcostalba, I've created an issue where I've tried to explain my point with an example position: #948.

**Pull request closed without merging**